File: frontend/src/lib/api/registration.js
// Client for registration, email verification, password reset, and TOTP
// enrollment (internal/auth/loginui). Unlike loginSession.js, none of these
// calls need an in-progress ZITADEL session — registration and password
// reset happen before one exists.
import { LoginSessionError } from "./loginSession.js"

/**
 * @param {string} path
 * @param {unknown} body
 */
async function post(path, body) {
  const response = await fetch(path, {
    method: "POST",
    credentials: "same-origin",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
  const payload = response.status === 204 ? {} : await response.json().catch(() => ({}))
  if (!response.ok) {
    throw new LoginSessionError(payload.error || `request to ${path} failed`, response.status)
  }
  return payload
}

/**
 * Creates a new account and emails a verification link. Returns
 * `{ userId }`.
 * @param {{loginName: string, email: string, password: string, givenName: string, familyName: string}} fields
 */
export function register(fields) {
  return post("/auth/register", fields)
}

/**
 * Confirms a registration's emailed verification code.
 * @param {string} userId
 * @param {string} code
 */
export function verifyEmail(userId, code) {
  return post("/auth/verify-email", { userId, code })
}

/**
 * Emails a password-reset link for loginName, if an account exists. Always
 * succeeds regardless of whether the account exists.
 * @param {string} loginName
 */
export function requestPasswordReset(loginName) {
  return post("/auth/password-reset", { loginName })
}

/**
 * Sets a new password using an emailed reset code.
 * @param {string} userId
 * @param {string} code
 * @param {string} newPassword
 */
export function resetPassword(userId, code, newPassword) {
  return post("/auth/password-reset/confirm", { userId, code, newPassword })
}

/**
 * Begins TOTP enrollment for userId. Returns `{ uri, secret }`.
 * @param {string} userId
 */
export function startTOTPEnrollment(userId) {
  return post("/auth/mfa/totp", { userId })
}

/**
 * Completes TOTP enrollment.
 * @param {string} userId
 * @param {string} code
 */
export function confirmTOTPEnrollment(userId, code) {
  return post("/auth/mfa/totp/verify", { userId, code })
}

File: internal/auth/loginui/registration.go
package loginui

import (
	"context"
	"errors"
	"fmt"
)

// Register creates a new ZITADEL user and emails a verification link. It
// returns the ZITADEL user ID: not a secret (verifying still requires the
// emailed code), and needed by the frontend to complete verification and
// optional TOTP enrollment afterward.
func (s *Service) Register(
	ctx context.Context, loginName, email, password, givenName, familyName string,
) (string, error) {
	userID, emailCode, err := s.adminAPI.registerHuman(ctx, loginName, email, password, givenName, familyName)
	if err != nil {
		return "", err
	}
	link := fmt.Sprintf("%s/verify-email?userId=%s&code=%s", s.frontendBaseURL, userID, emailCode)
	body := fmt.Sprintf(
		"Welcome! Verify your email by visiting the link below.\n\n%s\n\n"+
			"If you didn't create this account, you can ignore this email.",
		link,
	)
	if err := s.mailer.Send(ctx, email, "Verify your email", body); err != nil {
		return "", fmt.Errorf("send verification email: %w", err)
	}
	return userID, nil
}

// VerifyEmail confirms a registration's email-verification code.
func (s *Service) VerifyEmail(ctx context.Context, userID, code string) error {
	return s.adminAPI.verifyEmail(ctx, userID, code)
}

// RequestPasswordReset emails a password-reset link for loginName if a
// matching account exists. It succeeds either way, so the caller can't use
// it to enumerate accounts.
func (s *Service) RequestPasswordReset(ctx context.Context, loginName string) error {
	userID, email, err := s.adminAPI.findUserByLoginName(ctx, loginName)
	if err != nil {
		if errors.Is(err, errInvalidCredentials) {
			return nil // unknown login name: succeed anyway to avoid leaking account existence
		}
		return err
	}
	code, err := s.adminAPI.requestPasswordReset(ctx, userID)
	if err != nil {
		return err
	}
	link := fmt.Sprintf("%s/reset-password?userId=%s&code=%s", s.frontendBaseURL, userID, code)
	body := fmt.Sprintf(
		"Reset your password by visiting the link below.\n\n%s\n\n"+
			"If you didn't request this, you can ignore this email.",
		link,
	)
	if err := s.mailer.Send(ctx, email, "Reset your password", body); err != nil {
		return fmt.Errorf("send password reset email: %w", err)
	}
	return nil
}

// ResetPassword sets a new password for userID using the code from
// RequestPasswordReset's email.
func (s *Service) ResetPassword(ctx context.Context, userID, code, newPassword string) error {
	return s.adminAPI.setPassword(ctx, userID, newPassword, code)
}

// StartTOTPEnrollment begins TOTP enrollment for userID (called right after
// registration, as an optional "enable 2FA now" step), returning the
// otpauth: URI and raw secret for the user's authenticator app.
func (s *Service) StartTOTPEnrollment(ctx context.Context, userID string) (string, string, error) {
	return s.adminAPI.registerTOTP(ctx, userID)
}

// ConfirmTOTPEnrollment completes TOTP enrollment for userID, verifying the
// caller controls the enrolled authenticator.
func (s *Service) ConfirmTOTPEnrollment(ctx context.Context, userID, code string) error {
	return s.adminAPI.verifyTOTPRegistration(ctx, userID, code)
}
